### PR TITLE
[Core] `usb_device_state`: consolidate usb state handling across implementations

### PR DESCRIPTION
**Pull request by KarlK90**
## Description

Previously all usb drivers and platform implementations (expect for our oddball atsam) tracked the same two global variables:
    
 - `keyboard_protocol`: to indicate if we are in report or boot protocol
 - `keyboard_idle`: for the idle_rate of the keyboard endpoint
    
And a local variable that was exposed trough some indirection:
   
 - `keyboard_led_state`: for the currently set indicator leds (caps lock etc.)
    
These have all been moved into the `usb_device_state` struct wich is accessible by getters and setters.
    
This reduces code duplication and centralizes the state management across platforms and drivers.

This is then put into use for usb hid specification section 7.2.6:
    
> When initialized, all devices default to report protocol. However the
> host should not make any assumptions about the device’s state and should
> set the desired protocol whenever initializing a device.
    
Thus on reset we should always do exactly that.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
action_util.c and report.c error due to not having USB_PROTOCOL_* defined.

**Comment by KarlK90** [comment]
> action_util.c and report.c error due to not having USB_PROTOCOL_* defined.

Thanks! Fixed now.

File: keyboards/takashicompany/minidivide/keyboard.json [hunk 1]
```diff
@@ -32,8 +32,7 @@
             "rainbow_swirl": true,
             "rgb_test": true,
```
[the remark this answers is not in the document]
Comment (KarlK90) [reply keyboards/takashicompany/minidivide/keyboard.json]: True, applied.

**Comment by drashna** [comment]
Looks like there are some lint errors for OS detection, and the tests are failing for it, too.

**Comment by KarlK90** [comment]
Thanks, I've addressed the failing tests.

**Review by drashna — APPROVED** [review]

**Pull request merged**